js/source/tile_pyramid.js

```diff
@@ -146,22 +146,36 @@ TilePyramid.prototype = {
      * @private
      */
     findLoadedChildren: function(coord, maxCoveringZoom, retain) {
-        var complete = true;
-        var z = coord.z;
-        var coords = coord.children(this.maxzoom);
-        for (var i = 0; i < coords.length; i++) {
-            var id = coords[i].id;
-            if (this._tiles[id] && this._tiles[id].loaded) {
-                retain[id] = true;
-            } else {
-                complete = false;
-                if (z < maxCoveringZoom) {
-                    // Go further down the hierarchy to find more unloaded children.
-                    this.findLoadedChildren(coords[i], maxCoveringZoom, retain);
+        var found = false;
+
+        for (var id in this._tiles) {
+            var tile = this._tiles[id];
+
+            // only consider loaded tiles on higher zoom levels
+            if (retain[id] || !tile.loaded || tile.coord.z <= coord.z) continue;
+
+            // disregard tiles that are not descendants of the given tile coordinate
+            var z2 = Math.pow(2, Math.min(tile.coord.z, this.maxzoom) - Math.min(coord.z, this.maxzoom));
+            if (Math.floor(tile.coord.x / z2) !== coord.x ||
+                Math.floor(tile.coord.y / z2) !== coord.y)
+                continue;
+
+            // found loaded child
+            retain[id] = true;
+            found = true;
+
+            // loop through parents; retain the topmost loaded one if found
+            while (tile && tile.coord.z - 1 > coord.z) {
+                var parentId = tile.coord.parent(this.maxzoom).id;
+                tile = this._tiles[parentId];
+
+                if (tile && tile.loaded) {
+                    delete retain[id];
+                    retain[parentId] = true;
                 }
             }
         }
-        return complete;
+        return found;
```

Comment:
This changes the sense of the return value: before it was "did we find children that completely cover the parent tile", now it's "did we find one or more children". The return value is used in both contexts where `findLoadedChildren` is called, but I'm not sure how important the existing semantics are. What do you think?